lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Shared.hs

```diff
@@ -0,0 +1,217 @@
+                                 State
+-------------------------------------------------------------------------------}
+
+-- | A state to keep track of script templates, account public keys for cosigners,
+-- | and a combination of possible script addresses to look for in a ledger
+data SharedState (n :: NetworkDiscriminant) k = SharedState
+    { accountKey :: k 'AccountK XPub
+        -- ^ Reward account public key associated with this wallet
+    , derivationPrefix :: DerivationPrefix
+        -- ^ Derivation path prefix from a root key up to the account key
+    , derivationKeyNumber :: KeyNumberScope
+        -- ^ Number of first keys that are used to produce candidate addresses
+    , paymentScriptTemplate :: !(Maybe ScriptTemplate)
+        -- ^ Script template together with a map of account keys and cosigners
+        -- for payment credential
+    , stakingScriptTemplate :: !(Maybe ScriptTemplate)
+        -- ^ Script template together with a map of account keys and cosigners
+        -- for staking credential
+    , addressCandidates :: [Address]
+    }
+    deriving stock (Generic)
+
+deriving instance
+    ( Show (k 'AccountK XPub)
+    ) => Show (SharedState n k)
+
+deriving instance
+    ( Eq (k 'AccountK XPub)
+    ) => Eq (SharedState n k)
+
+instance
+    ( NFData (k 'AccountK XPub)
+    )
+    => NFData (SharedState n k)
+
+newtype KeyNumberScope =
+    KeyNumberScope { unKeyNumberScope :: Word8 }
+    deriving (Eq, Generic, Show)
+    deriving anyclass NFData
+
+-- | Construct a SharedState for a wallet from public account key and its corresponding index,
+-- script templates for both staking and spending, and the number of keys used for
+-- generating the corresponding address candidates.
+mkSharedState
+    :: forall (n :: NetworkDiscriminant) k. WalletKey k
+    => k 'AccountK XPub
+    -> Index 'Hardened 'AccountK
+    -> Maybe ScriptTemplate
+    -> Maybe ScriptTemplate
+    -> KeyNumberScope
+    -> Maybe (SharedState n k)
+mkSharedState accXPub accIx spendingTemplate stakingTemplate keyNum =
+    let
+        prefix =
+            DerivationPrefix ( purposeBIP44, coinTypeAda, accIx )
+        accXPub' = liftXPub $ getRawKey accXPub
+        rewardXPub =
+            getKey $ deriveAddressPublicKey accXPub' Stake minBound
+        addressesToFollow = case (spendingTemplate, stakingTemplate) of
+            (Nothing, Nothing) -> []
+            (Just template', Nothing) ->
+                if (isRight $ validateScriptTemplate RecommendedValidation template') then
+                    generateAddressCombination template' rewardXPub keyNum
+                else
+                    []
+            _ -> undefined
+    in
+        if all isNothing [spendingTemplate, stakingTemplate] then
+            Nothing
+        else
+            Just $ SharedState accXPub prefix keyNum spendingTemplate stakingTemplate addressesToFollow
+
+replaceCosignersWithVerKeys
+    :: ScriptTemplate
+    -> Map Cosigner Word8
```

Comment:
I am not sure to get this type? The derivation index should be the same for all co-signers (by design).